crates/solver/src/s3_instance_upload.rs

```diff
@@ -0,0 +1,112 @@
+use anyhow::Result;
+use aws_sdk_s3::{types::ByteStream, Client, Credentials, Region};
+use aws_types::credentials::SharedCredentialsProvider;
+use chrono::{DateTime, Utc};
+use model::auction::AuctionId;
+
+#[derive(Default)]
+pub struct Config {
+    pub region: String,
+    pub bucket: String,
+    pub access_key_id: String,
+    pub secret_access_key: String,
+    /// Prepended to the auction id to form the final filename. Something like
+    /// "staging/mainnet/quasimodo/". Should end with `/` if intended to be a folder.
+    pub filename_prefix: String,
+}
+
+pub struct S3InstanceUploader {
+    bucket: String,
+    filename_prefix: String,
+    client: Client,
+}
+
+impl S3InstanceUploader {
+    pub fn new(config: Config) -> Self {
+        let aws_config = aws_types::sdk_config::Builder::default()
+            .region(Region::new(config.region))
+            .credentials_provider(SharedCredentialsProvider::new(Credentials::from_keys(
+                config.access_key_id,
+                config.secret_access_key,
+                None,
+            )))
+            .build();
+        Self {
+            bucket: config.bucket,
+            filename_prefix: config.filename_prefix,
+            client: Client::new(&aws_config),
+        }
+    }
+
+    /// Upload the bytes (expected to represent a json encoded solver instance) to the configured S3
+    /// bucket.
+    ///
+    /// The final filename is the configured prefix followed by `{current_date}/{auction_id}`.
+    pub async fn upload_instance(&self, auction: AuctionId, value: Vec<u8>) -> Result<()> {
+        let key = self.filename(chrono::offset::Utc::now(), auction);
+        self.upload(key, value).await
+    }
+
+    fn filename(&self, now: DateTime<Utc>, auction: AuctionId) -> String {
+        let date = now.format("%Y-%m-%d");
+        format!("{}{date}/{auction}.json", self.filename_prefix)
+    }
+
+    async fn upload(&self, key: String, value: Vec<u8>) -> Result<()> {
+        self.client
```

Comment:
Depending on how expensive storage in s3 is we could compress the file here (ideally dependent on a CLI flag).
Judging by a single test we could save:
* ~90% disk space with `gzip`
* ~91.6% with `brotli`
The downside would be that you would have to decompress the files to know their content not sure how much of an inconvenience would be.

Reply:
According to Anuj we should pay at most ~40$ per month at the moment. However, since we launched limit orders the S3 disk space obviously grows more rapidly than before:
![Screenshot 2023-01-05 at 1 29 37 PM](https://user-images.githubusercontent.com/19190235/210731046-8a1eeb17-6473-4aa0-9c6f-c9b3e6277515.png)

Reply:
Compression is also on my radar. It makes using the instances slightly less convenient because you have to decompress them locally first but I also think saving 90% is worth it.